src/QMCHamiltonians/StressPBC.cpp

```diff
@@ -0,0 +1,349 @@
+  SpeciesSet& tspeciesB(P.getSpeciesSet());
+  int ChargeAttribIndxA = tspeciesA.addAttribute("charge");
+  int MemberAttribIndxA = tspeciesA.addAttribute("membersize");
+  int ChargeAttribIndxB = tspeciesB.addAttribute("charge");
+  int MemberAttribIndxB = tspeciesB.addAttribute("membersize");
+  NptclA                = PtclA.getTotalNum();
+  NptclB                = P.getTotalNum();
+  NumSpeciesA           = tspeciesA.TotalNum;
+  NumSpeciesB           = tspeciesB.TotalNum;
+  //Store information about charges and number of each species
+  Zat.resize(NptclA);
+  Zspec.resize(NumSpeciesA);
+  Qat.resize(NptclB);
+  Qspec.resize(NumSpeciesB);
+  NofSpeciesA.resize(NumSpeciesA);
+  NofSpeciesB.resize(NumSpeciesB);
+  for (int spec = 0; spec < NumSpeciesA; spec++)
+  {
+    Zspec[spec]       = tspeciesA(ChargeAttribIndxA, spec);
+    NofSpeciesA[spec] = static_cast<int>(tspeciesA(MemberAttribIndxA, spec));
+  }
+  for (int spec = 0; spec < NumSpeciesB; spec++)
+  {
+    Qspec[spec]       = tspeciesB(ChargeAttribIndxB, spec);
+    NofSpeciesB[spec] = static_cast<int>(tspeciesB(MemberAttribIndxB, spec));
+  }
+
+  for (int spec = 0; spec < NumSpeciesA; spec++) {}
+  RealType totQ = 0.0;
+  for (int iat = 0; iat < NptclA; iat++)
+    totQ += Zat[iat] = Zspec[PtclA.GroupID[iat]];
+  for (int iat = 0; iat < NptclB; iat++)
+    totQ += Qat[iat] = Qspec[P.GroupID[iat]];
+
+  AA = LRCoulombSingleton::getDerivHandler(P);
+}
+
+SymTensor<StressPBC::RealType, OHMMS_DIM> StressPBC::evaluateLR_AB(ParticleSet& P)
+{
+  SymTensor<RealType, OHMMS_DIM> res = 0.0;
+  const StructFact& RhoKA(*(PtclA.SK));
+  const StructFact& RhoKB(*(P.SK));
+
+  for (int i = 0; i < NumSpeciesA; i++)
+  {
+    SymTensor<RealType, OHMMS_DIM> esum;
+    esum = 0.0;
+    for (int j = 0; j < NumSpeciesB; j++)
+    {
+#if defined(USE_REAL_STRUCT_FACTOR)
+      esum += Qspec[j] *
+          AA->evaluateStress(RhoKA.KLists.kshell, RhoKA.rhok_r[i], RhoKA.rhok_i[i], RhoKB.rhok_r[j], RhoKB.rhok_i[j]);
+#else
+      esum += Qspec[j] * AA->evaluateStress(RhoKA.KLists.kshell, RhoKA.rhok[i], RhoKB.rhok[j]);
+
+#endif
+    }
+    res += Zspec[i] * esum;
+  }
+ 
+  return res;
+}
+
+SymTensor<StressPBC::RealType, OHMMS_DIM> StressPBC::evaluateSR_AB(ParticleSet& P)
+{
+  const auto& d_ab = P.getDistTable(ei_table_index);
+  SymTensor<RealType, OHMMS_DIM> res = 0.0;
+  //Loop over distinct eln-ion pairs
+  for (int jpart = 0; jpart < NptclB; jpart++)
+  {
+    const auto& drijs = d_ab.getDisplRow(jpart);
+    const auto& rijs = d_ab.getDistRow(jpart);
+    const RealType e = Qat[jpart];
+    for (int iat = 0; iat < NptclA; iat++)
+    {
+      res += Zat[iat]*e*AA->evaluateSR_dstrain(drijs[iat], rijs[iat]);
+    }
+  }
+  return res;
+}
+
+SymTensor<StressPBC::RealType, OHMMS_DIM> StressPBC::evaluateSR_AA(ParticleSet& P)
+{ // !!!! HACK to obtain AA distance table using firstTimeStress
+  int itabSelf = ee_table_index;
+  if (firstTimeStress) itabSelf = ii_table_index;
```

Comment:
It is better to pass the table index by an argument explicitly than relying on a state firstTimeStress.

Reply:
OK. Addressed.

Reply:
@Paul-St-Young Probably the rationale is obvious, but the use of unnecessary use of state machines or reliance on particular state information has proven to be one of the major pain points with the existing code (understanding, maintenance, source of bugs, accessibility to new folks...). So  if you find yourself writing new code, please avoid relying on state wherever you can.

Reply:
@prckent good point. Will avoid the state machine as much as possible in the future.